[aotinductor] Support additional lifted constants supplied to const folding.

zhxchen17 opened the pull request:
Summary:
In export workflow, we always have a lifted graph which doesn't fetch constants through get_attr nodes. This cause some compatibility issue when we're trying to use inductor's split_const_gm function with a lifted graph.

This diff make an additive change to split_const_gm's interface, such that, when the pass sees a placeholder node is present in the lifted_constants table, it will also use that as the source of constness.

This change won't break the existing code and the lifted_constants table can be used orthogonal to the existing const folding mechanisms.

Also as required from MTIA team, we want to introduce a small callback function used to skip certain nodes during const folding.

For the internal followup counterpart, see D59685145

Test Plan: buck run mode/opt caffe2/test:test_export -- -r split_const_gm

Differential Revision: D59692790


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

desertfire reviewed (commented):
Is this PR ready for review? I am still seeing CI failures.

While I agree making AOTI work on lifted graphs is a good direction, but don't we always unlift graphs in AOTI at the moment, https://github.com/pytorch/pytorch/blob/0851de5b167d3856285ad2a370c0355453d328dc/torch/_inductor/compile_fx.py#L1521C23-L1521C36 ?

zhxchen17 commented:
> Is this PR ready for review? I am still seeing CI failures.
> 
> While I agree making AOTI work on lifted graphs is a good direction, but don't we always unlift graphs in AOTI at the moment, https://github.com/pytorch/pytorch/blob/0851de5b167d3856285ad2a370c0355453d328dc/torch/_inductor/compile_fx.py#L1521C23-L1521C36 ?

Thanks the the comment! @desertfire yeah this is ready for review. the test failures are due to different Python version and not related to the core changes made here. I will fix them shortly.

> but don't we always unlift graphs in AOTI at the moment

Also I'm not trying to impose the unlifting requirement onto aoti, as mentioned in the diff summary, after the change this pass can work with both lifted and unlifted graphs.

SherlockNoMad commented on torch/_inductor/constant_folding.py:
```diff
@@ -253,23 +276,41 @@ def constant_graph_tag(gm: torch.fx.GraphModule):
 
 
-def run_and_get_constant_graph(gm: torch.fx.GraphModule) -> torch.fx.GraphModule:
+def run_and_get_constant_graph(
```
Does this supports gm with a mixed of get_attr and lifted constants? 
or do we not expect ppl to have a graph?

zhxchen17 replied:
yeah we support mixture of nodes. the invariant is simpler to specify this way. we should always have a graph.

zhxchen17 committed:
[aotinductor] Support additional lifted constants supplied to const folding. (#130743)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/130743

In export workflow, we always have a lifted graph which doesn't fetch constants through get_attr nodes. This cause some compatibility issue when we're trying to use inductor's split_const_gm function with a lifted graph.

This diff make an additive change to split_const_gm's interface, such that, when the pass sees a placeholder node is present in the lifted_constants table, it will also use that as the source of constness.

This change won't break the existing code and the lifted_constants table can be used orthogonal to the existing const folding mechanisms.

Also as required from MTIA team, we want to introduce a small callback function used to skip certain nodes during const folding.

For the internal followup counterpart, see D59685145

Test Plan: buck run mode/opt caffe2/test:test_export -- -r split_const_gm

Reviewed By: SherlockNoMad

Differential Revision: D59692790
